src/client/activeobjectmgr.cpp

```diff
@@ -25,28 +25,33 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 namespace client
 {
 
-void ActiveObjectMgr::clear()
+ActiveObjectMgr::~ActiveObjectMgr()
 {
-	// delete active objects
-	for (auto &active_object : m_active_objects) {
-		delete active_object.second;
-		// Object must be marked as gone when children try to detach
-		active_object.second = nullptr;
+	if (!m_active_objects.empty()) {
+		warningstream << "client::ActiveObjectMgr::~ActiveObjectMgr(): not cleared."
+				<< std::endl;
+		clear();
 	}
-	m_active_objects.clear();
 }
 
 void ActiveObjectMgr::step(
 		float dtime, const std::function<void(ClientActiveObject *)> &f)
 {
 	g_profiler->avg("ActiveObjectMgr: CAO count [#]", m_active_objects.size());
-	for (auto &ao_it : m_active_objects) {
-		f(ao_it.second);
+
+	// Same as in server activeobjectmgr.
+	std::vector<u16> ids = getAllIds();
+
+	for (u16 id : ids) {
+		auto it = m_active_objects.find(id);
+		if (it == m_active_objects.end())
+			continue; // obj was removed
+		f(it->second.get());
```

Comment:
from an efficiency POV this is quite bad
the loop goes from `O(n)` to `O(n * log n)` just for lookup

if we set entries to `nullptr` and postpone deletion wouldn't that solve the problem? IIRC it was just about iterator invalidation

Reply:
The currently iterated obj isn't necessarily deleted here, and neither in `clearIf`.

>if we set entries to `nullptr` and postpone deletion

Do you mean in `removeObject`? That's an interesting idea.
`f` can also insert new objects. Insertion doesn't invalidate iterators, but it means some of the new objects are immediately iterated through here and others aren't. So I'm not sure about the correctness.

IIRC, these functions are only called once per client or server step. So until this really turns out to be a performance bottleneck, I'd rather opt for the current more naive solution than to attempting premature optimization.